python/lsst/pipe/tasks/hips.py

```diff
@@ -604,3 +614,907 @@ def build_quantum_graph(
+            Do color pngs instead of per-band grayscale.
+        """
+        min_order = self.config.min_order
+
+        if not do_color:
+            png_grayscale_mapping = AsinhMapping(
+                self.config.png_gray_asinh_minimum,
+                self.config.png_gray_asinh_stretch,
+                Q=self.config.png_gray_asinh_softening,
+            )
+        else:
+            png_color_mapping = AsinhMapping(
+                self.config.png_color_asinh_minimum,
+                self.config.png_color_asinh_stretch,
+                Q=self.config.png_color_asinh_softening,
+            )
+
+            bcb = self.config.blue_channel_band
+            gcb = self.config.green_channel_band
+            rcb = self.config.red_channel_band
+            colorstr = f'{bcb}{gcb}{rcb}'
+
+        # The base path is the hips_base_uri with the band as a sub-directory.
+        hips_base_path = ResourcePath(self.config.hips_base_uri, forceDirectory=True)
+
+        # We need to unique-ify the pixels because they show up for multiple bands.
+        # The output of this is a sorted array.
+        pixels = np.unique(np.array([pixel for pixel, _ in hips_exposure_handle_dict.keys()]))
+
+        # Add a "gutter" pixel at the end.  Start with 0 which maps to 0 always.
+        pixels = np.append(pixels, [0])
+
+        # Convert the pixels to each order that will be generated
+        pixels_shifted = {}
+        pixels_shifted[max_order] = pixels
+        for order in range(max_order - 1, min_order - 1, -1):
+            pixels_shifted[order] = np.right_shift(pixels_shifted[order + 1], 2)
+
+        # And set the gutter to an illegal pixel value
+        for order in range(min_order, max_order + 1):
+            pixels_shifted[order][-1] = -1
+
+        # Read in the first pixel for determining image properties
+        exp0 = hips_exposure_handle_dict[list(hips_exposure_handle_dict.keys())[0]].get()
+        bbox = exp0.getBBox()
+        npix = bbox.getWidth()
+        shift_order = int(np.round(np.log2(npix)))
+
+        # Create blank exposures for each level, including the highest order
+        # We also make sure we create blank exposures for any bands used in the color
+        # pngs, even if they aren't available.
+        exposures = {}
+        for band in bands:
+            for order in range(min_order, max_order + 1):
+                exp = exp0.Factory(bbox=bbox)
+                exp.image.array[:, :] = np.nan
+                exposures[(band, order)] = exp
+
+        # Loop over all pixels, avoiding the gutter.
+        for pixel_counter, pixel in enumerate(pixels[:-1]):
+            self.log.debug("Working on high resolution pixel %d", pixel)
+            for band in bands:
+                # Read all the exposures here for the highest order.
+                if (pixel, band) in hips_exposure_handle_dict:
+                    exposures[(band, max_order)] = hips_exposure_handle_dict[(pixel, band)].get()
```

Comment:
`band, max_order` vs `pixel, band`  ordering?

Is this max_order by definition because the healpix generation code only generates images at the highest order? Therefore it's not possible `pixel != max_order`. Can we have a comment for that?

Reply:
I added this comment:
```
                # There will always be at least one band with a HiPS image available
                # at the highest order. However, for color images it is possible that
                # not all bands have coverage so we require this check.
```